### PR TITLE
Add output from hydra-queue-runner for unrunnable steps (no machine).

### PR DESCRIPTION
**Pull request by kquick**
This helps generate log information (from the hydra-queue-runner service) about steps for which there is no available machine.  This is currently a silent failure.

This patch works by seeing if any of the currently busy machines could build the remaining steps and if a step cannot be built by any busy machine (and it wasn't already assigned to a non-busy machine) then it reports it as a step with no machine available to build it.

This might over-report in the case where build machines where down, but I think that is still a valid condition for reporting.

There is no Web-UI display of this condition; it is found by examining the log output from the hydra-queue-runner service.  Web-UI display would need to have a web-page replication of this logic, or else an addition to the database to store this information, both of which are more ambitious undertakings left to (possibly) future pull request efforts.

Issues:
  * https://github.com/NixOS/hydra/issues/568
  * https://github.com/NixOS/hydra/issues/289
  * https://github.com/NixOS/hydra/issues/502
  * https://github.com/NixOS/hydra/issues/266
  * the tail of https://github.com/NixOS/hydra/issues/433
  * possibly helpful for https://github.com/NixOS/hydra/issues/317

### CODE REVIEW

**Comment by kquick** [comment]
Sample output:

`Apr 23 11:58:34 hydra hydra-queue-runner[28483[: NO MACHINE AVAILABLE to run step '/nix/store/b41lzs0xshmkq6bban48h8k4y207abxz-llvm-5.0.2.drv' (needs system type 'x86_64-linux:big-parallel')`

File: src/hydra-queue-runner/dispatcher.cc [hunk 1]
```diff
@@ -282,6 +287,19 @@ system_time State::doDispatch()
...
+        for (auto & stepInfo : runnableSorted) {
+            auto & step(stepInfo.step);
+            bool couldRunStep = false;
+            for (auto & mp : busyMachines)
+                if (mp->supportsStep(step)) {
+                    couldRunStep = true;
+                    break;
+                }
+            if (couldRunStep) continue;
+            printMsg(lvlError, format("NO MACHINE AVAILABLE to run step '%1%' (needs system type '%2%')") %
```
Comment (grahamc) [inline src/hydra-queue-runner/dispatcher.cc]:
```suggestion
            printMsg(lvlError, format("No machine available to run step '%1%' (needs system type '%2%')") %
```

**Comment by kquick** [comment]
Please hold off a merge... this is overreporting and needs an update (which I'm working on).

**Comment by knl** [comment]
> Please hold off a merge... this is overreporting and needs an update (which I'm working on).

Hi @kquick do you have an update on this one? We found it might be useful on our deployment, as we stumbled upon the same problem several times.

**Pull request closed without merging**